client/offchain/src/api/http.rs

```diff
@@ -715,6 +711,12 @@ mod tests {
 					let server = hyper::Server::bind(&"127.0.0.1:0".parse().unwrap())
 						.serve(hyper::service::make_service_fn(|_| { async move {
 							Ok::<_, Infallible>(hyper::service::service_fn(move |_req| async move {
+								// Take at least *a bit* of time so that we can test the non-blocking
+								// nature of the HTTP. Sometimes we receive the answer immediately,
+								// which means we can't know if we blockingly waited for a response
+								// or we did it in a non-blocking fashion but it arrived immediately.
+								tokio::time::delay_for(std::time::Duration::from_millis(10)).await;
```

Comment:
This will fail again. 

Reading this "Note" in the corresponding test already shows that this is broken. If this test does not expect any answer, you should modify `build_api_server!` to make the answer configurable, aka for this test there would be no answer.

Reply:
This is not limited to a single test in one instance - the note applies to every test where we assert that headers for a valid request is still empty due to the nature of the API.

As I understand it, making the answer configurable misses the point. We'll need to sprinkle every call site with the same dummy logic making it more verbose.

We don't need anything else from the endpoint to be a dead simple HTTP server. What's being 'broken' here is that we immediately ask for response headers in a non-blocking fashion and expect the answer to not arrive immediately.

We could not implicitly wait in `response_headers` or not test for empty headers in so many instances but making the answer configurable will not allow us anything here IIUC.

FWIW I can confirm that with this patch the test isn't flaky locally when executed for some time but can bump the delay if needed.

Reply:
Or could we not run the http server on the same thread and only `poll` when we expect it to process something? Does it needs to run in the background?

Reply:
Changed the internals of future deadline calculation so that we can immediately return on

https://github.com/paritytech/substrate/blob/74eebfa42e9b94d60814b322a2e1c44472785f13/client/offchain/src/api/http.rs#L299
 https://github.com/paritytech/substrate/blob/74eebfa42e9b94d60814b322a2e1c44472785f13/client/offchain/src/api/http.rs#L323-L340

Because we implicitly wait with a deadline of `timestamp::now`
https://github.com/paritytech/substrate/blob/74eebfa42e9b94d60814b322a2e1c44472785f13/client/offchain/src/api/http.rs#L396

So this behaves as expected as in we immediately return result with a zero timeout.